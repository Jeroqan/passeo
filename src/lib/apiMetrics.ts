import { NextRequest, NextResponse } from 'next/server';
// Vercel Edge Runtime'da prom-client çalışmadığı için metrics'i devre dışı bırakıyoruz
// import { httpRequests } from './metrics';

type ApiHandler = (
  req: NextRequest,
  context: { params: any }
) => Promise<Response | NextResponse>;

export function withApiMetrics(handler: ApiHandler) {
  return async (
    req: NextRequest,
    context: { params: any }
  ): Promise<Response | NextResponse> => {
    const { pathname } = req.nextUrl;
    let response: Response | NextResponse;
    try {
      response = await handler(req, context);
      // Vercel Edge Runtime'da metrics devre dışı
      // httpRequests.inc({
      //   method: req.method,
      //   route: pathname,
      //   status_code: response.status,
      // });
      return response;
    } catch (error) {
      // Vercel Edge Runtime'da metrics devre dışı
      // httpRequests.inc({
      //   method: req.method,
      //   route: pathname,
      //   status_code: 500,
      // });
      // Hatayı yeniden fırlatarak Next.js'in normal hata işleme mekanizmasının
      // çalışmasına izin veriyoruz.
      throw error;
    }
  };
} 