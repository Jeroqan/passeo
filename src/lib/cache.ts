import crypto from 'crypto';
import type { AiDetectionApiResponse } from './aiDetectionLogic';

// Define the structure of the data we'll be caching.
// This now directly uses the API response type for consistency.
export type DetectCacheData = AiDetectionApiResponse;

// Simple in-memory cache using a Map.
const detectCache = new Map<string, DetectCacheData>();

/**
 * Generates a SHA-256 hash for a given text.
 * @param text The text to hash.
 * @returns A hex-encoded SHA-256 hash.
 */
function hashText(text: string): string {
  return crypto.createHash('sha256').update(text, 'utf8').digest('hex');
}

export { detectCache, hashText }; 