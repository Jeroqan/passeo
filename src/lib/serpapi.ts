import { getJson } from 'serpapi';

if (!process.env.SERPAPI_API_KEY) {
  throw new Error('SERPAPI_API_KEY is not defined');
}

export async function searchKeywords(keyword: string) {
  const response = await getJson({
    api_key: process.env.SERPAPI_API_KEY,
    engine: 'google',
    q: keyword,
    google_domain: 'google.com.tr',
    gl: 'tr',
    hl: 'tr',
  });

  return response;
} 