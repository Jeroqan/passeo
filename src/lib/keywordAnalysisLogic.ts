import { getJson } from 'serpapi';
import OpenAI from 'openai';
import { buildKeywordExtractionPrompt } from '@/lib/promptBuilder/keywordPrompt';

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY,
});

// OpenAI kullanarak teknik özellikleri analiz eden yardımcı fonksiyon
const analyzeTechnicalFeatures = async (productName: string, category: string, brand?: string): Promise<string[]> => {
  try {
    const userPrompt = buildKeywordExtractionPrompt({ productName, brand, category });

    const completion = await openai.chat.completions.create({
      model: "gpt-4-turbo-preview",
      messages: [
        {
          role: "user",
          content: userPrompt
        }
      ],
      temperature: 0.2,
      max_tokens: 250,
    });

    const featuresString = completion.choices[0].message.content;
    if (!featuresString) {
      return [];
    }

    // Virgülle ayrılmış dizeyi diziye çevir ve gereksiz boşlukları temizle
    return featuresString.split(',').map(feature => feature.trim()).filter(feature => feature.length > 0);

  } catch (error) {
    console.error('OpenAI teknik özellik analizi hatası:', error);
    // Hata durumunda, hatayı yukarıya fırlat ki çağıran fonksiyon haberdar olsun.
    throw error;
  }
};

// SerpApi ile arama hacmi ve trend verilerini al
const getSearchData = async (query: string) => {
  try {
    const response = await getJson({
      api_key: process.env.SERPAPI_KEY,
      engine: 'google',
      q: query,
      google_domain: 'google.com.tr',
      gl: 'tr',
      hl: 'tr',
      num: 100
    });

    // Arama sonuçlarından ilgili terimleri çıkar
    const relatedTerms = response.related_searches?.map((item: any) => item.query) || [];
    
    // Arama hacmi ve trend verilerini hesapla
    const searchVolume = Math.floor(Math.random() * 1000) + 500; // Gerçek API'de bu veri gelecek
    const trend = Math.floor(Math.random() * 30) + 70; // Gerçek API'de bu veri gelecek

    return {
      searchVolume,
      trend,
      relatedTerms: relatedTerms.slice(0, 5)
    };
  } catch (error) {
    console.error('SerpApi hatası:', error);
    // Hata durumunda, hatayı yukarıya fırlat.
    throw error;
  }
};

interface ProductInfo {
    name: string;
    category: string;
    brand?: string;
}

export const analyzeKeywords = async (productInfo: ProductInfo, isFree: boolean) => {
    const { name: productName, brand, category } = productInfo;

    if (isFree) {
        const technicalFeatures = await analyzeTechnicalFeatures(productName, category, brand);
        
        const keywords = technicalFeatures.map((feature, i) => ({
            text: feature,
            trend: 80 - i * 5 + Math.floor(Math.random() * 10) - 5, // Dummy trend
            type: 'technical',
            searchVolume: 1000 - i * 50 + Math.floor(Math.random() * 100) - 50, // Dummy search volume
            relatedTerms: []
        }));
        return { keywords };
    }
    
    const technicalFeatures = await analyzeTechnicalFeatures(productName, category, brand);
    
    const keywordPromises = technicalFeatures.map(async (feature) => {
      const queryForSerpApi = `${productName} ${feature}`;
      const searchData = await getSearchData(queryForSerpApi);

      let type: 'technical' | 'feature' | 'specification' | 'comparison' = 'feature';
      if (feature.includes('işlemci') || feature.includes('ram') || feature.includes('depolama')) {
        type = 'specification';
      } else if (feature.includes('pil') || feature.includes('batarya')) {
        type = 'technical';
      }

      return {
        text: feature,
        type: type,
        searchVolume: searchData.searchVolume,
        trend: searchData.trend,
        relatedTerms: searchData.relatedTerms
      };
    });

    let keywords = await Promise.all(keywordPromises);

    const generalQuery = `${productName} teknik özellikleri`;
    const generalSearchData = await getSearchData(generalQuery);
    keywords.push({
      text: generalQuery,
      type: 'technical',
      searchVolume: generalSearchData.searchVolume,
      trend: generalSearchData.trend,
      relatedTerms: generalSearchData.relatedTerms
    });

    keywords.sort((a, b) => b.trend - a.trend);

    return { keywords };
} 