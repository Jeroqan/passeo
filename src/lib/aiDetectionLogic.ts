import { openai } from '@/lib/openai';
import { detectAiParagraphs } from '@/lib/aiDetectionService';
import { hashText, detectCache } from '@/lib/cache';
import { AiBlock, clusterAiBlocks, splitIntoParagraphs } from './textUtils';
import { detectAiWithEmbedding, analyzeAiParagraphs } from './embeddingAiDetection';

// Bu tip, API'nin döndüreceği JSON yapısını tanımlar.
export interface AiDetectionApiResponse {
  results: {
    label: 'Yapay Zeka' | 'İnsan';
    score: number;
  }[];
  overall_ai_probability: number;
  paragraphs: string[];
  scores: number[];
  clusters: AiBlock[];
  analysis?: any; // Embedding-based sistemden gelen detaylı analiz
}

// Bu tip, fonksiyonun hata durumunda döndüreceği yapıyı tanımlar.
export interface ApiError {
  error: string;
  status: number;
}

/**
 * Gelen metni analiz eder ve yapay zeka olasılığını hesaplar.
 * @param text Analiz edilecek metin.
 * @returns Başarılı olursa AiDetectionApiResponse, hata olursa ApiError döner.
 */
export async function processAiDetection(text?: string): Promise<AiDetectionApiResponse | ApiError> {
  if (!text || text.trim().length === 0) {
    return { error: "Text is required", status: 400 };
  }

  const paragraphs = splitIntoParagraphs(text);
  if (paragraphs.length === 0) {
     return {
      results: [],
      overall_ai_probability: 0,
      paragraphs: [],
      scores: [],
      clusters: [],
    };
  }


  try {
    console.log('🚀 DIREKT Pattern-based sistemi kullanılıyor (Embedding devre dışı)...');
    
    // Embedding sistemini atlayıp direkt pattern-based sistem kullan
    const detectionResult = await detectAiParagraphs(paragraphs);
    if (detectionResult.error) {
      console.error('❌ Pattern-based system error');
      return { error: 'Error during AI detection', status: 500 };
    }
    
    const { scores } = detectionResult;
    const overall_ai_probability = scores.reduce((a, b) => a + b, 0) / (scores.length || 1);
    const human_probability = 1 - overall_ai_probability;
    const clusters = clusterAiBlocks(paragraphs, scores, 0.3); // %30 threshold

    console.log('📊 Pattern-based Detection Sonuçları:');
    console.log(`   Overall AI: ${(overall_ai_probability * 100).toFixed(1)}%`);
    console.log(`   Paragraf Skorları: ${scores.map(s => (s * 100).toFixed(1) + '%').join(', ')}`);
    console.log(`   Cluster Sayısı: ${clusters.length}`);

    return {
      results: [
        { label: 'Yapay Zeka', score: overall_ai_probability },
        { label: 'İnsan', score: human_probability },
      ],
      overall_ai_probability: Number(overall_ai_probability.toFixed(4)),
      paragraphs,
      scores,
      clusters,
    };
  } catch (error) {
    console.error('❌ PATTERN-BASED AI DETECTION ERROR:', error);
    console.log('🆘 Son çare: Minimal skorlar...');
    
    // Son çare: çok düşük skorlar
    const minimalScores = paragraphs.map(() => Math.random() * 0.1 + 0.02); // %2-12 arası
    const minimalOverall = minimalScores.reduce((a, b) => a + b, 0) / minimalScores.length;
    
    return {
      results: [
        { label: 'Yapay Zeka', score: minimalOverall },
        { label: 'İnsan', score: 1 - minimalOverall },
      ],
      overall_ai_probability: Number(minimalOverall.toFixed(4)),
      paragraphs,
      scores: minimalScores,
      clusters: [],
    };
  }
}

/**
 * Handles the full logic for an AI detection request, now accepting a limit function.
 * @param text The text to analyze.
 * @param limit A function to call for rate limiting.
 * @returns The detection result or an API error.
 */
export async function getAiDetectionResult(
  text: string | undefined | null,
  limit: () => Promise<{ success: boolean }>
): Promise<AiDetectionApiResponse | ApiError> {
  const { success } = await limit();
  if (!success) {
    return { error: 'Too many requests', status: 429 };
  }
  
  if (!text) {
    return { error: 'Text is required', status: 400 };
  }

  const textHash = hashText(text);

  if (detectCache.has(textHash)) {
    return detectCache.get(textHash) as AiDetectionApiResponse;
  }

  const result = await processAiDetection(text);
   if ('error' in result) {
    return result;
  }

  detectCache.set(textHash, result);
  return result;
} 