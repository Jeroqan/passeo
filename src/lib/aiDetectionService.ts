import { openai } from '@/lib/openai';

interface DetectionResult {
  scores: number[];
  error?: boolean;
}

/**
 * AI detection patterns - Bu kalıplar AI tarafından üretilen metinlerde daha sık görülür
 */
const AI_PATTERNS = {
  // Çok teknik/formal dil kalıpları
  technical: [
    /\b(optimize|enhance|utilize|implement|facilitate|comprehensive|innovative|cutting-edge|state-of-the-art)\b/gi,
    /\b(furthermore|moreover|additionally|consequently|therefore|thus|hence)\b/gi,
    /\b(significantly|substantially|considerably|remarkably|notably)\b/gi,
  ],
  
  // Fazla mükemmel grammar kalıpları  
  perfect: [
    /\b(it is important to note that|it should be noted that|it is worth noting that)\b/gi,
    /\b(in order to|for the purpose of|with the aim of)\b/gi,
    /\b(a wide range of|a variety of|a multitude of|numerous|various)\b/gi,
  ],
  
  // Türkçe AI kalıpları
  turkish: [
    /\b(bu bağlamda|bu çerçevede|bu doğrultuda|bu nedenle|bu şekilde)\b/gi,
    /\b(özellikle|genellikle|çoğunlukla|büyük ölçüde)\b/gi,
    /\b(gelişmiş|yenilikçi|modern|teknolojik|profesyonel)\b/gi,
    /\b(sağlamak|sunmak|oluşturmak|geliştirmek|tasarlamak)\b/gi,
  ],

  // Listeleme kalıpları (AI çok sever)
  listing: [
    /^\s*\d+\.\s/gm, // 1. 2. 3. numaralı listeler
    /^\s*[-•]\s/gm,  // bullet point'ler
    /\b(ilk olarak|ikinci olarak|üçüncü olarak|son olarak)\b/gi,
  ],

  // Çok uzun cümleler (AI eğilimi)
  longSentences: /[.!?]\s+[A-ZÜĞIÖÇŞ][^.!?]{100,}/g,
  
  // Aşırı detaylı açıklamalar
  overExplaining: [
    /\b(detaylı|ayrıntılı|kapsamlı|eksiksiz|tam|bütün)\b/gi,
    /\b(her türlü|tüm|bütün|tamamen|mükemmel)\b/gi,
  ]
};

/**
 * Paragraf bazında AI detection skoru hesapla - ULTRA optimize edilmiş versiyon
 * Hedef: %5-15 arası skorlar, insan metinleri için çok düşük sonuçlar
 */
function calculateAiScore(text: string): number {
  if (!text || text.trim().length < 10) return 0;
  
  let score = 0.005; // Base score çok düşürüldü (2% → 0.5%)
  const words = text.split(/\s+/).length;
  
  // Pattern matching - minimal penaltı
  for (const [category, patterns] of Object.entries(AI_PATTERNS)) {
    if (category === 'longSentences') continue; // Ayrı handle edilecek
    
    if (Array.isArray(patterns)) {
      patterns.forEach((pattern: RegExp) => {
        const matches = text.match(pattern) || [];
        score += matches.length * 0.02; // Her match +2% (5% → 2%)
      });
    }
  }
  
  // Uzun cümle kontrolü - çok minimal penaltı
  const longSentences = text.match(AI_PATTERNS.longSentences) || [];
  score += longSentences.length * 0.03; // Her uzun cümle +3% (8% → 3%)
  
  // Kelime sayısı bazlı skor - minimal
  if (words > 80) score += 0.01; // Threshold yükseltildi: 50→80, 3% → 1%
  if (words > 120) score += 0.01; // Threshold yükseltildi: 100→120, 3% → 1%
  if (words > 200) score += 0.01; // Threshold yükseltildi: 150→200, 3% → 1%
  
  // Çok teknik kelime yoğunluğu - çok yüksek threshold
  const technicalWords = (text.match(/\b(teknoloji|sistem|özellik|işlev|performans|kalite|verimlilik)\b/gi) || []).length;
  if (technicalWords / words > 0.12) score += 0.04; // Threshold: 0.08 → 0.12, Penaltı: 0.08 → 0.04
  
  // Çok mükemmel yazım - çok yüksek threshold
  const punctuationDensity = (text.match(/[,.;:!?]/g) || []).length / words;
  if (punctuationDensity > 0.25) score += 0.02; // Threshold: 0.20 → 0.25, Penaltı: 0.04 → 0.02
  
  // Fazla formal kelimeler - çok yüksek threshold
  const formalWords = (text.match(/\b(ancak|fakat|lakin|ama|çünkü|zira|dolayısıyla)\b/gi) || []).length;
  if (formalWords / words > 0.08) score += 0.03; // Threshold: 0.05 → 0.08, Penaltı: 0.06 → 0.03
  
  // İnsan sinyalleri - dengeli negatif etki
  const humanSignals = (text.match(/\b(çok|gerçekten|bence|sanırım|galiba|herhalde|muhtemelen|belki|ya|valla|yani|şey|böyle|işte)\b/gi) || []).length;
  if (humanSignals > 0) score -= humanSignals * 0.03; // Her insan sinyali için -3%
  
  // Konuşma dili ifadeleri - dengeli negatif
  const conversationalPatterns = (text.match(/\b(bunu|şunu|kesinlikle|açıkçası|doğrusu|honestly|genel olarak|aldım|kullanıyorum|memnunum)\b/gi) || []).length;
  if (conversationalPatterns > 0) score -= conversationalPatterns * 0.04; // Her konuşma ifadesi için -4%
  
  // Birinci şahıs ifadeleri (güçlü insan sinyali)
  const firstPersonPatterns = (text.match(/\b(ben|benim|bana|benden|bende|aldım|kullanıyorum|düşünüyorum|seviyorum|beğendim)\b/gi) || []).length;
  if (firstPersonPatterns > 0) score -= firstPersonPatterns * 0.05; // Her birinci şahıs için -5%
  
  // Duygusal ifadeler (AI nadir kullanır)
  const emotionalPatterns = (text.match(/\b(harika|süper|mükemmel|berbat|rezalet|muhteşem|fena değil|idare eder)\b/gi) || []).length;
  if (emotionalPatterns > 0) score -= emotionalPatterns * 0.03; // Her duygusal ifade için -3%
  
  // Base score'u biraz yükselt ve minimum'u ayarla
  score += 0.04; // Base boost +4%
  
  // Score'u gerçekçi aralıkta sınırla
  return Math.min(Math.max(score, 0.03), 0.35); // Minimum 3%, maksimum 35%
}

/**
 * Given a list of paragraphs, returns a confidence score for each.
 * @param paragraphs Array of texts to be analyzed.
 * @returns An object containing an array of confidence scores.
 */
export async function detectAiParagraphs(paragraphs: string[]): Promise<DetectionResult> {
  if (!paragraphs || paragraphs.length === 0) {
    return { scores: [] };
  }

  try {
    // Her paragraf için AI skoru hesapla
    const scores = paragraphs.map(paragraph => {
      if (paragraph.trim().length < 10) return 0;
      
      const aiScore = calculateAiScore(paragraph);
      console.log(`Paragraph: "${paragraph.substring(0, 50)}..." → AI Score: ${aiScore.toFixed(3)}`);
      
      return aiScore;
    });

    console.log('AI Detection Results:', scores);
    
    return {
      scores: scores,
    };
  } catch (e: any) {
    console.error('DETECT_AI_TEXT_ERROR:', e.message);
    return {
      scores: paragraphs.map(() => 0),
      error: true,
    };
  }
} 