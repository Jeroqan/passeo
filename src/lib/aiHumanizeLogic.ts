import { openai } from '@/lib/openai';
import { buildHumanizePrompt } from '@/lib/promptBuilder/humanizePrompt';

export interface HumanizeApiResponse {
  humanizedText: string;
}

export interface ApiError {
  error: string;
  status: number;
}

// Helper function to escape special characters for regex
function escapeRegExp(string: string) {
    return string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

export async function processHumanizeRequest(
  fullText: string | undefined | null,
  selected: string[] | undefined | null
): Promise<HumanizeApiResponse | ApiError> {
  if (!fullText || !selected || !Array.isArray(selected) || selected.length === 0) {
    return { error: 'Tam metin ve seçili paragraflar gereklidir.', status: 400 };
  }

  try {
    // Extract the title part of the text, which should not be rewritten
    const lines = fullText.split('\\n');
    let title = '';
    for (let i = 0; i < lines.length; i++) {
        if (lines[i].trim() !== '') {
            title = lines[i];
            break;
        }
    }

    const humanizePrompt = buildHumanizePrompt({ fullText, selected });

    const response = await openai.chat.completions.create({
      model: 'gpt-4o',
      messages: [{ role: 'user', content: humanizePrompt }],
      temperature: 0.7,
    });

    const humanizedBody = response.choices[0].message.content?.trim();

    if (!humanizedBody) {
      return { error: 'Modelden geçerli bir yanıt alınamadı.', status: 500 };
    }
    
    // Re-attach the original title to the humanized body
    const humanizedText = [title, humanizedBody].filter(Boolean).join('\\n\\n');
    return { humanizedText };

  } catch (error) {
    console.error('PROCESS_HUMANIZE_ERROR:', error);
    return { error: 'Dahili Sunucu Hatası', status: 500 };
  }
} 