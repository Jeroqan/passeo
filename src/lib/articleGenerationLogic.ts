import { openai } from '@/lib/openai';
import { z } from 'zod';
import { createArticlePrompt } from './promptBuilder/articlePrompt';
import { CoreMessage } from 'ai';

interface ProductInfo {
  name: string;
  brand: string;
  category: string;
}

interface ArticlePayload {
  productInfo: ProductInfo;
  keywords: string[];
  categoryUrl: string;
}

export interface ArticleApiResponse {
  content: string;
}

export interface ApiError {
  error: string;
  status: number;
}

export const ArticleGenerationSchema = z.object({
  productName: z.string().min(3, 'Ürün adı en az 3 karakter olmalıdır.'),
  brand: z.string().min(2, 'Marka en az 2 karakter olmalıdır.'),
  category: z.string().min(3, 'Kategori en az 3 karakter olmalıdır.'),
  keyword: z.string().min(3, 'Anahtar kelime en az 3 karakter olmalıdır.'),
});

export type ArticleGenerationParams = z.infer<typeof ArticleGenerationSchema>;

/**
 * Gelen makale üretme isteğini doğrular ve OpenAI'ye stream çağrısı yapar.
 * @param params - Makale üretimi için gerekli parametreler.
 * @returns Başarılı olursa OpenAI stream nesnesini, hata olursa ApiError döner.
 */
export async function processArticleGeneration(
  params: Partial<ArticleGenerationParams>
): Promise<ReturnType<typeof openai.chat.completions.create> | ApiError> {
  const validationResult = ArticleGenerationSchema.safeParse(params);
  if (!validationResult.success) {
    throw new Error(validationResult.error.errors.map(e => e.message).join(', '));
  }

  const { productName, brand, category, keyword } = validationResult.data;

  if (keyword.length > 100) {
    return {
      error: 'Anahtar kelime çok uzun.',
      status: 400,
    };
  }
  
  try {
    const prompt = createArticlePrompt({ productName, brand, category, keyword });

    const stream = await openai.chat.completions.create({
      model: 'gpt-4o',
      messages: [
        { role: 'system', content: 'Sen profesyonel bir teknik içerik yazarı ve SEO uzmanısın. Kullanıcıya sadece HTML formatında, <body> etiketleri olmadan doğrudan makale içeriğini döndür.' },
        { role: 'user', content: prompt }
      ],
      temperature: 0.7,
      stream: true,
    });

    return stream;
  } catch (error) {
    console.error('[ARTICLE GEN LOGIC ERROR]', error);
    return {
      error: 'Makale üretimi sırasında bir hata oluştu.',
      status: 500,
    };
  }
}

export async function generateArticleStream(params: ArticleGenerationParams) {
  const validationResult = ArticleGenerationSchema.safeParse(params);
  if (!validationResult.success) {
    const errorMessage = validationResult.error.errors.map(e => e.message).join(', ');
    throw new Error(errorMessage);
  }

  const { productName, brand, category, keyword } = validationResult.data;

  const prompt = createArticlePrompt({ productName, brand, category, keyword });

  const stream = await openai.chat.completions.create({
    model: 'gpt-4o',
    stream: true,
    messages: [
      {
        role: 'system',
        content: "You are a helpful assistant that provides responses in HTML format, chunk by chunk.",
      },
      {
        role: 'user',
        content: prompt,
      },
    ],
  });

  return stream;
}

export function prepareArticleGeneration(params: ArticleGenerationParams): { messages: CoreMessage[] } {
  const validationResult = ArticleGenerationSchema.safeParse(params);
  if (!validationResult.success) {
    const errorMessage = validationResult.error.errors.map(e => e.message).join(', ');
    throw new Error(errorMessage);
  }

  const { productName, brand, category, keyword } = validationResult.data;
  const prompt = createArticlePrompt({ productName, brand, category, keyword });

  const messages: CoreMessage[] = [
    {
      role: 'system',
      content: "You are a helpful assistant that provides responses in HTML format, chunk by chunk.",
    },
    {
      role: 'user',
      content: prompt,
    },
  ];

  return { messages };
} 