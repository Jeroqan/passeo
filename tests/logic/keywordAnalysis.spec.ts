import { analyzeKeywords } from '@/lib/keywordAnalysisLogic';
import { getJson } from 'serpapi';
import OpenAI from 'openai';

// Manuel mock'u etkinleştir
jest.mock('openai');
jest.mock('serpapi');

// Manuel mock'tan mockCreate fonksiyonunu al ve tipi düzelt
const { mockCreate } = OpenAI as any;

const mockedGetJson = getJson as jest.Mock;

describe('analyzeKeywords Logic', () => {
  const productInfo = {
    name: 'Test Product',
    category: 'Test Category',
    brand: 'Test Brand',
  };

  beforeEach(() => {
    // Her testten önce mock'ları temizle
    jest.clearAllMocks();
    mockCreate.mockClear();
  });

  describe('Free Mode (isFree = true)', () => {
    it('should return keywords based on OpenAI features without calling SerpApi', async () => {
      // OpenAI mock'unu ayarla
      const mockFeatures = 'özellik 1, özellik 2, özellik 3';
      mockCreate.mockResolvedValue({
        choices: [{ message: { content: mockFeatures } }],
      });

      const result = await analyzeKeywords(productInfo, true);

      // Sonucu doğrula
      expect(result.keywords).toBeInstanceOf(Array);
      expect(result.keywords.length).toBe(3);
      expect(result.keywords[0].text).toBe('özellik 1');
      expect(mockedGetJson).not.toHaveBeenCalled();
    });

    it('should throw an error if OpenAI fails', async () => {
        mockCreate.mockRejectedValue(new Error('OpenAI API Error'));
        await expect(analyzeKeywords(productInfo, true)).rejects.toThrow('OpenAI API Error');
    });
  });

  describe('Normal Mode (isFree = false)', () => {
    it('should return keywords using both OpenAI and SerpApi, sorted by trend', async () => {
      const mockFeatures = 'pil ömrü, kamera kalitesi';
      mockCreate.mockResolvedValue({
        choices: [{ message: { content: mockFeatures } }],
      });

      mockedGetJson.mockImplementation((params: { q: string }) => {
        if (params.q === 'Test Product pil ömrü') {
          return Promise.resolve({ trend: 85, searchVolume: 1000, related_searches: [] });
        }
        if (params.q === 'Test Product kamera kalitesi') {
          return Promise.resolve({ trend: 95, searchVolume: 1500, related_searches: [] });
        }
        if (params.q === 'Test Product teknik özellikleri') {
          return Promise.resolve({ trend: 75, searchVolume: 500, related_searches: [] });
        }
        return Promise.resolve({ trend: 0, searchVolume: 0, related_searches: [] });
      });

      const result = await analyzeKeywords(productInfo, false);

      // Sonucun anlık görüntüsüyle (snapshot) eşleştiğini doğrula
      expect(result.keywords).toMatchSnapshot();
    });

    it('should throw an error if SerpApi fails', async () => {
        mockCreate.mockResolvedValue({ choices: [{ message: { content: 'feature1' } }] });
        mockedGetJson.mockRejectedValue(new Error('SerpApi Error'));
        await expect(analyzeKeywords(productInfo, false)).rejects.toThrow('SerpApi Error');
    });
  });
}); 