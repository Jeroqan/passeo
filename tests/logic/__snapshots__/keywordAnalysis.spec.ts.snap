// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`analyzeKeywords Logic Normal Mode (isFree = false) should return keywords using both OpenAI and SerpApi, sorted by trend 1`] = `
[
  {
    "relatedTerms": [],
    "searchVolume": 1158,
    "text": "kamera kalitesi",
    "trend": 94,
    "type": "feature",
  },
  {
    "relatedTerms": [],
    "searchVolume": 759,
    "text": "Test Product teknik özellikleri",
    "trend": 79,
    "type": "technical",
  },
  {
    "relatedTerms": [],
    "searchVolume": 1082,
    "text": "pil ömrü",
    "trend": 76,
    "type": "technical",
  },
]
`;
